File: platforms/android/module/src/main/kotlin/rs/whisker/runtime/ElementBinding.kt
package rs.whisker.runtime

/** Intrinsic-measurement policy negotiated with the Rust runtime. */
public enum class WhiskerMeasurement { None, Text, ReplacedContent, Custom }

/** Rust-owned child semantics. Native mount targets remain Host-local. */
public enum class WhiskerChildPolicy { None, Elements, PlainText }

/** Top-level shape of a value carried by [WhiskerValue]. */
public enum class WhiskerValueKind { Null, Bool, Int, Float, String, Bytes, Array, Map }

/** Runtime-assigned property identity received during surface bootstrap. */
public data class WhiskerPropertyBinding(
    public val id: Int,
    public val name: String,
    public val value: WhiskerValueKind,
) {
    init {
        require(id > 0) { "property id must be non-zero" }
        require(name.isNotEmpty()) { "property name must not be empty" }
    }
}

/** Runtime-assigned event identity received during surface bootstrap. */
public data class WhiskerEventBinding(
    public val id: Int,
    public val name: String,
    public val detail: WhiskerValueKind? = null,
) {
    init {
        require(id > 0) { "event id must be non-zero" }
        require(name.isNotEmpty()) { "event name must not be empty" }
    }
}

/** Runtime-assigned command identity received during surface bootstrap. */
public data class WhiskerCommandBinding(
    public val id: Int,
    public val name: String,
    public val arguments: WhiskerValueKind,
) {
    init {
        require(id > 0) { "command id must be non-zero" }
        require(name.isNotEmpty()) { "command name must not be empty" }
    }
}

/**
 * Rust-owned element registration decoded from a snapshot frame.
 *
 * This is runtime negotiation data, not generated Kotlin source. Host module
 * authors declare names in `ModuleDefinition`; the registry resolves those
 * names to these compact IDs once, before applying frame operations.
 */
public class WhiskerElementRegistration(
    public val elementType: Int,
    public val name: String,
    public val childPolicy: WhiskerChildPolicy,
    public val measurement: WhiskerMeasurement,
    public val properties: List<WhiskerPropertyBinding> = emptyList(),
    public val events: List<WhiskerEventBinding> = emptyList(),
    public val commands: List<WhiskerCommandBinding> = emptyList(),
) {
    private val propertiesById = properties.associateBy { it.id }
    private val eventsById = events.associateBy { it.id }
    private val commandsById = commands.associateBy { it.id }

    init {
        require(elementType > 0) { "element type must be non-zero" }
        require(name.isNotEmpty() && '@' !in name) {
            "element name must be non-empty and versionless"
        }
        require(propertiesById.size == properties.size) { "duplicate property id in $name" }
        require(properties.map { it.name }.toSet().size == properties.size) {
            "duplicate property name in $name"
        }
        require(eventsById.size == events.size) { "duplicate event id in $name" }
        require(events.map { it.name }.toSet().size == events.size) {
            "duplicate event name in $name"
        }
        require(commandsById.size == commands.size) { "duplicate command id in $name" }
        require(commands.map { it.name }.toSet().size == commands.size) {
            "duplicate command name in $name"
        }
    }

    public fun property(id: Int): WhiskerPropertyBinding =
        requireNotNull(propertiesById[id]) { "unknown property id $id on $name" }

    public fun event(id: Int): WhiskerEventBinding =
        requireNotNull(eventsById[id]) { "unknown event id $id on $name" }

    public fun command(id: Int): WhiskerCommandBinding =
        requireNotNull(commandsById[id]) { "unknown command id $id on $name" }
}

File: platforms/ios/Sources/WhiskerModule/ElementBinding.swift
import Foundation

/// Intrinsic-measurement policy negotiated with the Rust runtime.
public enum WhiskerMeasurement: Hashable {
    case none, text, replacedContent, custom
}

/// Rust-owned child semantics. Native mount targets remain Host-local.
public enum WhiskerChildPolicy: Hashable {
    case none, elements, plainText

    public var acceptsElements: Bool { self == .elements }
    public var acceptsPlainText: Bool { self == .plainText }
}

/// Top-level shape of a value carried by `WhiskerValue`.
public enum WhiskerValueKind: Hashable {
    case null, bool, int, float, string, bytes, array, map
}

/// Runtime-assigned property identity received during surface bootstrap.
public struct WhiskerPropertyBinding: Hashable {
    public let id: Int
    public let name: String
    public let value: WhiskerValueKind

    public init(id: Int, name: String, value: WhiskerValueKind) {
        precondition(id > 0 && !name.isEmpty)
        self.id = id
        self.name = name
        self.value = value
    }
}

/// Runtime-assigned event identity received during surface bootstrap.
public struct WhiskerEventBinding: Hashable {
    public let id: Int
    public let name: String
    public let detail: WhiskerValueKind?

    public init(id: Int, name: String, detail: WhiskerValueKind? = nil) {
        precondition(id > 0 && !name.isEmpty)
        self.id = id
        self.name = name
        self.detail = detail
    }
}

/// Runtime-assigned command identity received during surface bootstrap.
public struct WhiskerCommandBinding: Hashable {
    public let id: Int
    public let name: String
    public let arguments: WhiskerValueKind

    public init(id: Int, name: String, arguments: WhiskerValueKind) {
        precondition(id > 0 && !name.isEmpty)
        self.id = id
        self.name = name
        self.arguments = arguments
    }
}

/// Rust-owned runtime registration, not generated Swift source.
public struct WhiskerElementRegistration {
    public let elementType: Int
    public let name: String
    public let childPolicy: WhiskerChildPolicy
    public let measurement: WhiskerMeasurement
    public let textStyle: Bool
    public let properties: [WhiskerPropertyBinding]
    public let events: [WhiskerEventBinding]
    public let commands: [WhiskerCommandBinding]

    public init(
        elementType: Int,
        name: String,
        childPolicy: WhiskerChildPolicy,
        measurement: WhiskerMeasurement,
        textStyle: Bool = false,
        properties: [WhiskerPropertyBinding] = [],
        events: [WhiskerEventBinding] = [],
        commands: [WhiskerCommandBinding] = []
    ) {
        precondition(elementType > 0)
        precondition(!name.isEmpty && !name.contains("@"))
        precondition(Set(properties.map(\.id)).count == properties.count)
        precondition(Set(properties.map(\.name)).count == properties.count)
        precondition(Set(events.map(\.id)).count == events.count)
        precondition(Set(events.map(\.name)).count == events.count)
        precondition(Set(commands.map(\.id)).count == commands.count)
        precondition(Set(commands.map(\.name)).count == commands.count)
        self.elementType = elementType
        self.name = name
        self.childPolicy = childPolicy
        self.measurement = measurement
        self.textStyle = textStyle
        self.properties = properties
        self.events = events
        self.commands = commands
    }
}
